source/common/tracing/zipkin_tracer_impl.cc

```diff
@@ -0,0 +1,194 @@
+  }
+  return false;
+}
+
+ZipkinDriver::TlsZipkinTracer::TlsZipkinTracer(Zipkin::Tracer&& tracer, ZipkinDriver& driver)
+    : tracer_(new Zipkin::Tracer(std::move(tracer))), driver_(driver) {}
+
+ZipkinDriver::ZipkinDriver(const Json::Object& config, Upstream::ClusterManager& cluster_manager,
+                           Stats::Store& stats, ThreadLocal::Instance& tls,
+                           Runtime::Loader& runtime, const LocalInfo::LocalInfo& local_info,
+                           Runtime::RandomGenerator& random_generator)
+    : cm_(cluster_manager),
+      tracer_stats_{ZIPKIN_TRACER_STATS(POOL_COUNTER_PREFIX(stats, "tracing.zipkin."))}, tls_(tls),
+      runtime_(runtime), local_info_(local_info), tls_slot_(tls.allocateSlot()) {
+
+  Upstream::ThreadLocalCluster* cluster = cm_.get(config.getString("collector_cluster"));
+  if (!cluster) {
+    throw EnvoyException(fmt::format("{} collector cluster is not defined on cluster manager level",
+                                     config.getString("collector_cluster")));
+  }
+  cluster_ = cluster->info();
+
+  if (cluster_->features() & Upstream::ClusterInfo::Features::HTTP2) {
+    throw EnvoyException(
+        fmt::format("Zipkin collector service (cluster {}) can be accessed over http1.1 only",
+                    cluster_->name()));
+  }
+
+  const std::string collector_endpoint = config.getString("collector_endpoint");
+
+  tls_.set(tls_slot_, [this, collector_endpoint, &random_generator](Event::Dispatcher& dispatcher)
+                          -> ThreadLocal::ThreadLocalObjectSharedPtr {
+                            Zipkin::Tracer tracer(local_info_.clusterName(), local_info_.address());
+                            tracer.setReporter(ZipkinReporter::NewInstance(
+                                std::ref(*this), std::ref(dispatcher), collector_endpoint));
+                            Runtime::RandomGeneratorPtr rand_ptr(&random_generator);
+                            tracer.setRandomGenerator(std::move(rand_ptr));
+                            return ThreadLocal::ThreadLocalObjectSharedPtr{
+                                new TlsZipkinTracer(std::move(tracer), *this)};
+                          });
+}
+
+SpanPtr ZipkinDriver::startSpan(Http::HeaderMap& request_headers, const std::string&,
+                                SystemTime start_time) {
+  Zipkin::Tracer& tracer = *tls_.getTyped<TlsZipkinTracer>(tls_slot_).tracer_;
+  ZipkinSpanPtr active_span;
+  Zipkin::SpanPtr new_zipkin_span;
+
+  if (request_headers.OtSpanContext()) {
+    // Get the open-tracing span context.
+    // This header contains a span's parent-child relationships set by the downstream Envoy.
+    // The context built from this header allows the Zipkin tracer to
+    // properly set the span id and the parent span id.
+    Zipkin::SpanContext context;
+
+    context.populateFromString(request_headers.OtSpanContext()->value().c_str());
+
+    // Create either a child or a shared-context Zipkin span.
```

Comment:
can you put a comment on when a child is created vs when shared-context Zipkin span.